PB-755: Spec for collection assets

benschs opened the pull request:
Add CRUD endpoints for collection level assets.
Add endpoints for collection asset upload management. As process for upload is the same as for feature assets, only link to description.

boecklic reviewed (approved):
:+1:

asteiner-swisstopo commented on spec/transaction/components/responses.yaml:
```diff
@@ -18,6 +25,16 @@ components:
+    CollectionAsset:
+      description: >-
+        The response is a document consisting of one asset of the collection.
+      headers:
+        ETag:
+          $ref: "../../components/headers.yaml#/components/headers/ETag"
```
For my understanding: Why does the `CollectionAssets` have no `headers` while `CollectionAsset` does?

benschs replied:
AFAIK the ETag header is only used if a single entity is returned, and does not really make sense if lists are returned.

asteiner-swisstopo commented on spec/transaction/paths.yaml:
```diff
@@ -789,6 +942,249 @@ paths:
+        - $ref: "../components/parameters.yaml#/components/parameters/IfNoneMatch"
+      responses:
+        "200":
+          description: Collection asset multipart upload description.
```
Looks like a stub description that has not been changed yet.

asteiner-swisstopo replied:
Now it is changed to

> Specified collection asset multipart upload.

I am not sure I understand what this means. Shouldn't it be something like "Did X successfully"? For example, `completeCollectionAssetMultipartUpload` has description "Asset multipart upload completed successfully."

asteiner-swisstopo replied:
Ok, fine with me :+1:

asteiner-swisstopo commented on spec/transaction/paths.yaml:
```diff
@@ -789,6 +942,249 @@ paths:
+        "404":
+          $ref: "../components/responses.yaml#/components/responses/NotFound"
+        "409":
+          description: Another upload is already in progress, you need first to complete or abort it before creating a new upload.
```
Shouldn't this be "you first need to complete" instead of "you need first to complete"?

asteiner-swisstopo replied:
Fixed here, thanks. Would you mind correcting also on line 660 where this is probably copied from?

asteiner-swisstopo commented on spec/transaction/paths.yaml:
```diff
@@ -789,6 +942,249 @@ paths:
+        A file part must be at least 5 MB except for the last one and at most 5 GB, otherwise the complete operation will fail.
+
+        *Note: in order to provide integrity check during the upload, the base64-encoded 128-bit MD5 digest of each part must be
+        computed and passed in the create endpoint. Then this digest must also be passed as `Content-MD5` header during the upload.*
```
Typo: "passed in" --> "passed to"

asteiner-swisstopo commented on spec/transaction/paths.yaml:
```diff
@@ -789,6 +942,249 @@ paths:
+        - Collection Asset Upload Management
+      summary: Create a new collection asset multipart upload
+      description: |
+        Create collection asset multipart upload.
+
+        A file part must be at least 5 MB except for the last one and at most 5 GB, otherwise the complete operation will fail.
+
+        *Note: in order to provide integrity check during the upload, the base64-encoded 128-bit MD5 digest of each part must be
+        computed and passed in the create endpoint. Then this digest must also be passed as `Content-MD5` header during the upload.*
```
As mentioned by @boecklic, we might want make it explicit that uploading parts can happen in parallel. Would this be the right place?

asteiner-swisstopo replied:
Ok thanks. Now mentioned in tags.yaml under tag "Asset Upload Management":
> You may also upload multiple parts in parallel.

asteiner-swisstopo commented on spec/transaction/paths.yaml:
```diff
@@ -562,6 +714,7 @@ paths:
     put:
       tags:
         - Asset Upload Management
+        - Collection Asset Upload Management
```
For my understanding: Is this tagged with "Collection Asset Upload Management" because you can also directly upload a part to a collection now, instead of only to an item?

asteiner-swisstopo replied:
Still open, would you mind clarifying please?

benschs replied:
Yes. This path is where the parts are uploaded to s3. I decided to add the tag so it shows up in both tag sections.
Depending on the use-case you can either upload a part of an asset, or of a collection asset.

asteiner-swisstopo reviewed (commented):
Hi @benschs,

As discussed, I have had a look and found mostly typos or unclear writing.

The only more important remarks are
- Should we mention the parallel upload capabilities?
- Is the response "500" intentional or should it be "5XX"?

Note to myself: `openapitransactional.yaml` is generated from the other files (`responses.yaml`, `paths.yaml`, `tags.yaml`).

benschs commented on spec/static/spec/v1/openapitransactional.yaml:
```diff
@@ -296,6 +296,16 @@ tags:
+    description: |
+      Collection Asset files are uploaded via the STAC API using the API requests described in this chapter.
+
+      The flow of the requests is the same as for assets that belong to features, which is described under [Asset Upload Management](#tag/Asset-Upload-Management)
```
I am not sure on the difference, it seems that feature comes from OGC, but STAC uses items. I use them more or less interchangeably.

benschs commented:
> Hi @benschs,
> 
> As discussed, I have had a look and found mostly typos or unclear writing.
> 
> The only more important remarks are
> 
>     * Should we mention the parallel upload capabilities?
> 
>     * Is the response "500" intentional or should it be "5XX"?
> 
> 
> Note to myself: `openapitransactional.yaml` is generated from the other files (`responses.yaml`, `paths.yaml`, `tags.yaml`).

Thanks for all the feedback, I think I fixed most of the typos.

I added a short sentence about parallel capabilities in the tag description for the multipart upload process.
I also unified the responses to only include "500", although we could even think about removing the 5XX completely from the spec, because AFAICT the user does not gain any information they could act upon.

asteiner-swisstopo commented on spec/transaction/paths.yaml:
```diff
@@ -107,6 +107,158 @@ paths:
+          $ref: "../components/responses.yaml#/components/responses/InvalidParameter"
+        "404":
+          $ref: "../components/responses.yaml#/components/responses/NotFound"
+        "500":
```
Changed to "500" everywhere for consistency.

asteiner-swisstopo commented on spec/transaction/paths.yaml:
```diff
@@ -789,6 +942,249 @@ paths:
+      description: >-
+        Complete the multipart upload process. After completion, the collection asset metadata are updated
+        with the new `file:checksum` from the upload and the parts are automatically deleted.
+        The Asset `href` field is also set if it was the first upload.
```
Typo: "assets" --> "asset's". Same on line 854.

asteiner-swisstopo commented:
> Thanks for all the feedback, I think I fixed most of the typos.
> 
> I added a short sentence about parallel capabilities in the tag description for the multipart upload process. I also unified the responses to only include "500", although we could even think about removing the 5XX completely from the spec, because AFAICT the user does not gain any information they could act upon.

Had another look. Most are fixed, some still open. Would you mind having another look please? :innocent:

benschs commented on spec/transaction/paths.yaml:
```diff
@@ -789,6 +942,249 @@ paths:
+        A file part must be at least 5 MB except for the last one and at most 5 GB, otherwise the complete operation will fail.
+
+        *Note: in order to provide integrity check during the upload, the base64-encoded 128-bit MD5 digest of each part must be
+        computed and passed in the create endpoint. Then this digest must also be passed as `Content-MD5` header during the upload.*
```
Thanks, I missed this. Should be done now. :)
